File: test/data/languages/dart/Inheritance.dart
/// Dart Inheritance test file for native extraction validation

// Simple abstract class
abstract class Animal {
  String name;

  Animal(this.name);

  String makeSound();
}

// Class extending another class
class Dog extends Animal {
  Dog(String name) : super(name);

  @override
  String makeSound() => 'Woof!';
}

// Mixin definition
mixin Loggable {
  void log(String message) {
    print('[LOG] $message');
  }
}

// Interface (implicit in Dart - any class can be an interface)
abstract class Runnable {
  void run();
}

// Class with extends and implements
class Pet extends Dog implements Runnable {
  Pet(String name) : super(name);

  @override
  void run() {
    print('$name is running...');
  }
}

// Class with extends, implements, and with (mixin)
class ServiceAnimal extends Dog with Loggable implements Runnable {
  ServiceAnimal(String name) : super(name);

  @override
  void run() {
    log('$name is running for service!');
  }
}

// Sealed class (Dart 3.0+)
sealed class Shape {
  double get area;
}

// Class extending sealed class
class Circle extends Shape {
  final double radius;
  Circle(this.radius);

  @override
  double get area => 3.14159 * radius * radius;
}

// Enum with mixin
enum DogBreed with Loggable {
  labrador,
  bulldog,
  poodle;

  void describe() {
    log('Breed: $name');
  }
}

// Extension type (Dart 3.3+)
extension type Meters(double value) {
  double toFeet() => value * 3.28084;
}

// Mixin with on clause
mixin Runner on Animal {
  void sprint() {
    print('$name is sprinting!');
  }
}

File: test/data/languages/java/Inheritance.java
package com.example.inheritance;

import java.io.Serializable;

/**
 * Java inheritance test file - demonstrates various inheritance patterns
 */

// Base class with no inheritance
abstract class Animal {
    protected String name;

    public Animal(String name) {
        this.name = name;
    }

    public abstract String speak();
}

// Single inheritance
class Dog extends Animal {
    public Dog(String name) {
        super(name);
    }

    @Override
    public String speak() {
        return "Woof!";
    }
}

// Single inheritance
class Cat extends Animal {
    public Cat(String name) {
        super(name);
    }

    @Override
    public String speak() {
        return "Meow!";
    }
}

// Interface declaration with extends
interface Runnable {
    void run();
}

// Interface extending another interface
interface Runner extends Runnable {
    void sprint();
}

// Interface with multiple extends
interface Athlete extends Runner, Serializable {
    void compete();
}

// Class with extends and implements
class Pet extends Dog implements Serializable {
    private String owner;

    public Pet(String name, String owner) {
        super(name);
        this.owner = owner;
    }
}

// Class with extends and multiple implements
class ServiceAnimal extends Dog implements Runnable, Serializable {
    public ServiceAnimal(String name) {
        super(name);
    }

    @Override
    public void run() {
        System.out.println(name + " is running");
    }
}

// Enum implementing an interface
enum DogBreed implements Serializable {
    LABRADOR,
    POODLE,
    BULLDOG
}

// Functional interface
@FunctionalInterface
interface SoundMaker {
    String makeSound();
}
